proc/groupby.go

```diff
@@ -46,9 +46,33 @@ func IsErrTooBig(err error) bool {
 const defaultGroupByLimit = 1000000
 
 func CompileGroupBy(node *ast.GroupByProc, zctx *resolver.Context) (*GroupByParams, error) {
+	astKeys := node.Keys
+
+	if node.InputSortDir != 0 && node.Duration.Seconds == 0 && len(astKeys) > 0 {
+		if _, ok := astKeys[0].Expr.(ast.FieldExpr); !ok {
+			return nil, fmt.Errorf("compiling groupby: cannot use -sorted in conjunction with a computed primary key")
+		}
+	}
+	if node.Duration.Seconds > 0 {
+		everyKey := ast.Assignment{
+			Target: "ts",
+			Expr: &ast.FunctionCall{
+				Node:     ast.Node{"FunctionCall"},
+				Function: "Time.trunc",
+				Args: []ast.Expression{
+					&ast.FieldRead{ast.Node{"FieldRead"}, "ts"},
+					&ast.Literal{ast.Node{"Literal"}, "int64", strconv.Itoa(node.Duration.Seconds)},
+				},
+			},
+		}
+		astKeys = append([]ast.Assignment{everyKey}, astKeys...)
+	}
+	if len(astKeys) == 0 {
+		return nil, fmt.Errorf("compiling groupby: must have at least one key (or -every)")
+	}
```

Comment:
I think it would be simpler to do this transformation in zql.peg instead of here. That'd let us drop ast.GroupByProc.Duration, too.

Reply:
(You mean in parser-support.go and parser-support.js, right?)

I had tried that early on but it wasn't simpler. More importantly, I don't think that parser code should be aware of specific zql functions like "Time.trunc", or doing this kind of transformation.  Ideally it would go in some intermediate compilation pass, but we just have one for the time being... so I put it there.

Reply:
(separately, and independently of this PR, Duration should be called Every).